lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php

```diff
@@ -1511,6 +1564,31 @@ protected function _validateAndCompleteAssociationMapping(array $mapping)
         // the sourceEntity.
         $mapping['sourceEntity'] = $this->name;
 
+        if (
+            PHP_VERSION_ID >= 70400
+            && isset($this->reflClass)
+            && $this->reflClass->hasProperty($mapping['fieldName'])
+        ) {
+            $property = $this->reflClass->getProperty($mapping['fieldName']);
```

Comment:
factor this whole block out as `$mapping = $this->_validateAndCompleteTypedAssociationMapping($mapping);`.

Reply:
Done